checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,284 @@
+
+        if (((SpecialBlock) b).getSpecialType() == SpecialBlock.SpecialBlockType.EXCEPTIONAL_EXIT) {
+          // Do nothing if the successor block is an exceptional exit point
+          continue;
+        }
+
+        findOwningFields(fieldsWithCMVs, cfgMethod.getMethod());
+      }
+
+      BlockWithFieldsFacts successor = new BlockWithFieldsFacts(b, fieldsWithCMVs);
+      if (visited.add(successor)) {
+        worklist.add(successor);
+      }
+    }
+  }
+
+  /**
+   * Checks all element in the fieldsWithCMVs and adds an {@code @Owning} annotation on a field if
+   * the called methods set of that field in {@link FieldWithCMVs} includes all the methods in
+   * the @MustCall type of the field. It also updates classToFieldToFinalizers with the new detected
+   * finalizer.
+   *
+   * @param fieldsWithCMVs the set of FieldWithCMVs
+   * @param methodTree the method tree
+   */
+  private void findOwningFields(Set<FieldWithCMVs> fieldsWithCMVs, Tree methodTree) {
+    for (FieldWithCMVs fieldsWithCMV : fieldsWithCMVs) {
+      List<String> mcValues = typeFactory.getMustCallValue(fieldsWithCMV.field);
+      if (fieldsWithCMV.calledMethods.containsAll(mcValues)) {
+        WholeProgramInference wpi = typeFactory.getWholeProgramInference();
+        if (wpi != null) {
+          wpi.addFieldDeclarationAnnotation(fieldsWithCMV.field, OWNING);
+          typeFactory.updateClassToFieldToFinalizers(
+              cfgMethod.getClassTree(), fieldsWithCMV.field, TreeUtils.elementFromTree(methodTree));
+        }
+      }
+    }
+  }
+
+  /**
+   * A pair of a {@link Block} and a set of FieldsWithCMV on entry to the block. Each FieldsWithCMV
+   * represents a set of fields with a least upper bound of methods that were called somewhere in
+   * the previous blocks.
+   */
+  private static class BlockWithFieldsFacts {
+
+    /** The block. */
+    public final Block block;
+
+    /** The set of FieldsWithCMV. */
+    public final ImmutableSet<FieldWithCMVs> fieldsWithCMV;
+
+    /**
+     * Create a new BlockWithFieldsFacts.
+     *
+     * @param b the block
+     * @param fieldsWithCMVs the set of FieldWithCMVs
+     */
+    public BlockWithFieldsFacts(Block b, Set<FieldWithCMVs> fieldsWithCMVs) {
+      this.block = b;
+      this.fieldsWithCMV = ImmutableSet.copyOf(fieldsWithCMVs);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      BlockWithFieldsFacts that = (BlockWithFieldsFacts) o;
+      return block.equals(that.block) && fieldsWithCMV.equals(that.fieldsWithCMV);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(block, fieldsWithCMV);
+    }
+  }
+
+  /**
+   * A pair of a field {@link Element} and a set of method names that were called on the field
+   * somewhere in a method body.
+   */
+  private static class FieldWithCMVs {
```

Comment:
Can we just name it `FieldAndCalledMethods`?  I got confused wondering what a CMV was.